File: Python/dal/borrow_service.py
from dal.db_helper import DBHelper
from datetime import datetime


class BorrowService:
    def __init__(self):
        self.db = DBHelper()

    def borrow_book(self, reader_id: int, book_id: int, borrow_date: str, due_date: str) -> bool:
        conn = self.db.get_connection()
        cursor = conn.cursor()
        try:
            cursor.execute("SELECT Status FROM Reader WHERE ReaderId=?", (reader_id,))
            reader_row = cursor.fetchone()
            if not reader_row or reader_row['Status'] != 0:
                return False

            cursor.execute("SELECT AvailableCount FROM Book WHERE BookId=?", (book_id,))
            book_row = cursor.fetchone()
            if not book_row or book_row['AvailableCount'] <= 0:
                return False

            cursor.execute("""
                INSERT INTO BorrowRecord (ReaderId, BookId, BorrowDate, DueDate, Status)
                VALUES (?, ?, ?, ?, 0)
            """, (reader_id, book_id, borrow_date, due_date))

            cursor.execute("UPDATE Book SET AvailableCount = AvailableCount - 1 WHERE BookId=?", (book_id,))

            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            print(f"借书失败：{e}")
            return False
        finally:
            conn.close()

    def get_reader_borrow_info(self, reader_id: int) -> list:
        sql = """
            SELECT br.BorrowId, br.BorrowDate, br.DueDate, b.BookName, b.ISBN, r.ReaderName, r.ReaderCategoryId
            FROM BorrowRecord br
            INNER JOIN Book b ON br.BookId = b.BookId
            INNER JOIN Reader r ON br.ReaderId = r.ReaderId
            WHERE br.ReaderId=? AND br.Status IN (0, 2)
            ORDER BY br.DueDate
        """
        rows, _ = self.db.execute_data_set(sql, (reader_id,))
        return rows

    def search_borrow_records(self, reader_id: int = None, isbn: str = None, borrow_id: int = None, status: int = None) -> list:
        sql = """
            SELECT br.BorrowId, br.BorrowDate, br.DueDate, br.ReturnDate, br.Status,
                   b.BookName, b.ISBN, r.ReaderName, r.ReaderId
            FROM BorrowRecord br
            INNER JOIN Book b ON br.BookId = b.BookId
            INNER JOIN Reader r ON br.ReaderId = r.ReaderId
            WHERE 1=1
        """
        params = []
        if status is not None:
            sql += " AND br.Status=?"
            params.append(status)
        else:
            sql += " AND br.Status IN (0, 2)"
        if reader_id:
            sql += " AND r.ReaderId=?"
            params.append(reader_id)
        if isbn:
            sql += " AND b.ISBN LIKE ?"
            params.append(f"%{isbn}%")
        if borrow_id:
            sql += " AND br.BorrowId=?"
            params.append(borrow_id)
        sql += " ORDER BY br.DueDate"
        rows, _ = self.db.execute_data_set(sql, tuple(params) if params else ())
        return rows

    def return_book(self, borrow_id: int, return_date: str, late_fee: float, remark: str) -> bool:
        conn = self.db.get_connection()
        cursor = conn.cursor()
        try:
            cursor.execute("SELECT BookId FROM BorrowRecord WHERE BorrowId=?", (borrow_id,))
            row = cursor.fetchone()
            if not row:
                return False
            book_id = row['BookId']

            cursor.execute("""
                UPDATE BorrowRecord
                SET ReturnDate=?, LateFee=?, Status=1, Remark=?
                WHERE BorrowId=?
            """, (return_date, late_fee, remark, borrow_id))

            cursor.execute("UPDATE Book SET AvailableCount = AvailableCount + 1 WHERE BookId=?", (book_id,))

            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            print(f"还书失败：{e}")
            return False
        finally:
            conn.close()

    def get_borrow_record_by_id(self, borrow_id: int) -> list:
        sql = """
            SELECT br.*, b.BookName, b.ISBN, r.ReaderName, r.ReaderCategoryId
            FROM BorrowRecord br
            INNER JOIN Book b ON br.BookId = b.BookId
            INNER JOIN Reader r ON br.ReaderId = r.ReaderId
            WHERE br.BorrowId=?
        """
        rows, _ = self.db.execute_data_set(sql, (borrow_id,))
        return rows

    def search_all_borrow_records(self, reader_id: int = None, reader_name: str = None,
                                   book_name: str = None, isbn: str = None,
                                   status: int = None, borrow_from: str = None,
                                   borrow_to: str = None, return_from: str = None,
                                   return_to: str = None, page_index: int = 1,
                                   page_size: int = 10) -> tuple:
        sql = """
            SELECT br.BorrowId, br.BorrowDate, br.DueDate, br.ReturnDate, br.Status, br.LateFee, br.Remark,
                   CASE
                       WHEN br.Status = 0 THEN '借出中'
                       WHEN br.Status = 1 THEN '已归还'
                       WHEN br.Status = 2 THEN '逾期'
                       ELSE '未知'
                   END as StatusText,
                   b.BookName, b.ISBN, b.BookId, r.ReaderName, r.ReaderId
            FROM BorrowRecord br
            INNER JOIN Book b ON br.BookId = b.BookId
            INNER JOIN Reader r ON br.ReaderId = r.ReaderId
            WHERE 1=1
        """
        params = []

        if reader_id is not None and reader_id != 0:
            sql += " AND r.ReaderId=?"
            params.append(reader_id)
        if reader_name:
            sql += " AND r.ReaderName LIKE ?"
            params.append(f"%{reader_name}%")
        if book_name:
            sql += " AND b.BookName LIKE ?"
            params.append(f"%{book_name}%")
        if isbn:
            sql += " AND b.ISBN LIKE ?"
            params.append(f"%{isbn}%")
        if status is not None:
            sql += " AND br.Status=?"
            params.append(status)
        if borrow_from:
            sql += " AND br.BorrowDate>=?"
            params.append(borrow_from)
        if borrow_to:
            sql += " AND br.BorrowDate<=?"
            params.append(borrow_to)
        if return_from:
            sql += " AND br.ReturnDate>=?"
            params.append(return_from)
        if return_to:
            sql += " AND br.ReturnDate<=?"
            params.append(return_to)

        count_sql = f"SELECT COUNT(*) FROM ({sql})"
        total = self.db.execute_scalar(count_sql, tuple(params))

        sql += " ORDER BY br.BorrowDate DESC"
        if page_size > 0:
            offset = (page_index - 1) * page_size
            sql += f" LIMIT {page_size} OFFSET {offset}"

        rows, columns = self.db.execute_data_set(sql, tuple(params))
        return rows, total

File: BM.DAL/BorrowService.cs
using System;
using System.Data;
using System.Data.SqlClient;
using BookManagement.DAL.Utility;
using BookManagement.Model;
namespace BookManagement.DAL
{
    public class BorrowService
    {
        /// <summary>
        /// 借书操作（需要事务处理）
        /// </summary>
        public bool BorrowBook(int readerId, int bookId, DateTime borrowDate, DateTime dueDate)
        {
            // 注意：这是一个需要事务的操作，涉及多个表的更新
            using (SqlConnection conn = new SqlConnection(SQLHelper.ConnectionString))
            {
                conn.Open();
                SqlTransaction transaction = conn.BeginTransaction();

                try
                {
                    // 1. 检查读者状态
                    string checkReaderSql = $"SELECT Status FROM Reader WHERE ReaderId = {readerId}";
                    SqlCommand cmdCheckReader = new SqlCommand(checkReaderSql, conn, transaction);
                    object readerStatus = cmdCheckReader.ExecuteScalar();
                    if (readerStatus == null || Convert.ToInt32(readerStatus) != 0) // 0:正常
                    {
                        transaction.Rollback();
                        return false;
                    }
                    string checkBookSql = $"SELECT AvailableCount FROM Book WHERE BookId = {bookId}";
                    SqlCommand cmdCheckBook = new SqlCommand(checkBookSql, conn, transaction);
                    object availableCount = cmdCheckBook.ExecuteScalar();
                    if (availableCount == null || Convert.ToInt32(availableCount) <= 0)
                    {
                        transaction.Rollback();
                        return false;
                    }

                    // 3. 插入借阅记录
                    string insertBorrowSql = $@"
                        INSERT INTO BorrowRecord (ReaderId, BookId, BorrowDate, DueDate, Status) 
                        VALUES ({readerId}, {bookId}, '{borrowDate:yyyy-MM-dd}', '{dueDate:yyyy-MM-dd}', 0)";
                    SqlCommand cmdInsertBorrow = new SqlCommand(insertBorrowSql, conn, transaction);
                    int rows = cmdInsertBorrow.ExecuteNonQuery();
                    if (rows <= 0)
                    {
                        transaction.Rollback();
                        return false;
                    }
                    string updateBookSql = $"UPDATE Book SET AvailableCount = AvailableCount - 1 WHERE BookId = {bookId}";
                    SqlCommand cmdUpdateBook = new SqlCommand(updateBookSql, conn, transaction);
                    rows = cmdUpdateBook.ExecuteNonQuery();
                    if (rows <= 0)
                    {
                        transaction.Rollback();
                        return false;
                    }

                    transaction.Commit();
                    return true;
                }
                catch (Exception)
                {
                    transaction.Rollback();
                    throw;
                }
            }
        }
        public DataTable GetReaderBorrowInfo(int readerId)
        {
            string sql = $@"
                SELECT 
                    br.BorrowId,  -- 注意：使用别名
                    br.BorrowDate, 
                    br.DueDate, 
                    b.BookName, 
                    b.ISBN,
                    r.ReaderName,
                    r.ReaderCategoryId
                FROM BorrowRecord br
                INNER JOIN Book b ON br.BookId = b.BookId
                INNER JOIN Reader r ON br.ReaderId = r.ReaderId
                WHERE br.ReaderId = {readerId} AND br.Status IN (0, 2)  -- 0:借出中, 2:逾期
                ORDER BY br.DueDate";
            DataSet ds = SQLHelper.ExecuteDataSet(sql);
            return ds?.Tables?[0];
        }
        public DataTable SearchBorrowRecords(int? readerId, string isbn, int? borrowId)
        {
            string sql = @"
                SELECT 
                    br.BorrowId,  -- 注意：使用别名
                    br.BorrowDate,
                    br.DueDate,
                    br.ReturnDate,
                    br.Status,
                    b.BookName,
                    b.ISBN,
                    r.ReaderName,
                    r.ReaderId
                FROM BorrowRecord br
                INNER JOIN Book b ON br.BookId = b.BookId
                INNER JOIN Reader r ON br.ReaderId = r.ReaderId
                WHERE br.Status IN (0, 2)";  // 0:借出中, 2:逾期

            if (readerId.HasValue)
                sql += $" AND r.ReaderId = {readerId.Value}";

            if (!string.IsNullOrEmpty(isbn))
                sql += $" AND b.ISBN LIKE '%{isbn}%'";

            if (borrowId.HasValue)
                sql += $" AND br.BorrowId = {borrowId.Value}";

            sql += " ORDER BY br.DueDate";

            DataSet ds = SQLHelper.ExecuteDataSet(sql);
            return ds?.Tables?[0];
        }
        public bool ReturnBook(int borrowId, DateTime returnDate, decimal lateFee, string remark)
        {
            using (SqlConnection conn = new SqlConnection(SQLHelper.ConnectionString))
            {
                conn.Open();
                SqlTransaction transaction = conn.BeginTransaction();

                try
                {
                    // 1. 获取图书ID
                    string getBookSql = $"SELECT BookId FROM BorrowRecord WHERE BorrowId= {borrowId}";
                    SqlCommand cmdGetBook = new SqlCommand(getBookSql, conn, transaction);
                    object bookIdObj = cmdGetBook.ExecuteScalar();

                    if (bookIdObj == null)
                    {
                        transaction.Rollback();
                        return false;
                    }

                    int bookId = Convert.ToInt32(bookIdObj); string updateBorrowSql = $@"
                        UPDATE BorrowRecord 
                        SET ReturnDate = '{returnDate:yyyy-MM-dd}', 
                            LateFee = {lateFee}, 
                            Status = 1,  -- 1:已归还
                            Remark = '{remark}'
                        WHERE BorrowId = {borrowId}";

                    SqlCommand cmdUpdateBorrow = new SqlCommand(updateBorrowSql, conn, transaction);
                    int rows = cmdUpdateBorrow.ExecuteNonQuery();

                    if (rows <= 0)
                    {
                        transaction.Rollback();
                        return false;
                    }

                    // 3. 更新图书库存
                    string updateBookSql = $"UPDATE Book SET AvailableCount = AvailableCount + 1 WHERE BookId = {bookId}";
                    SqlCommand cmdUpdateBook = new SqlCommand(updateBookSql, conn, transaction);
                    rows = cmdUpdateBook.ExecuteNonQuery();

                    if (rows <= 0)
                    {
                        transaction.Rollback();
                        return false;
                    }

                    transaction.Commit();
                    return true;
                }
                catch (Exception)
                {
                    transaction.Rollback();
                    throw;
                }
            }
        }
        public DataTable GetBorrowRecordById(int borrowId)
        {
            string sql = $@"
                SELECT 
                    br.*,
                    b.BookName,
                    b.ISBN,
                    r.ReaderName,
                    r.ReaderCategoryId
                FROM BorrowRecord br
                INNER JOIN Book b ON br.BookId = b.BookId
                INNER JOIN Reader r ON br.ReaderId = r.ReaderId
                WHERE br.BorrowId = {borrowId}";

            DataSet ds = SQLHelper.ExecuteDataSet(sql);
            return ds?.Tables?[0];
        }/// <summary>
         /// 多条件查询所有借阅记录（支持分页）
         /// </summary>
        public DataTable SearchAllBorrowRecords(int? readerId, string readerName, string bookName,
            string isbn, int? status,string category, DateTime? borrowFrom, DateTime? borrowTo,
            DateTime? returnFrom, DateTime? returnTo, int pageIndex, int pageSize, out int totalRecords)
        {
            totalRecords = 0;

            // 构建基础查询
            string sql = @"
        SELECT 
            br.BorrowId,
            br.BorrowDate,
            br.DueDate,
            br.ReturnDate,
            br.Status,
            CASE 
                WHEN br.Status = 0 THEN '借出中'
                WHEN br.Status = 1 THEN '已归还'
                WHEN br.Status = 2 THEN '逾期'
                ELSE '未知'
            END as StatusText,
            b.BookName,
            b.ISBN,
            b.BookId,
            r.ReaderName,
            r.ReaderId
        FROM BorrowRecord br
        INNER JOIN Book b ON br.BookId = b.BookId
        INNER JOIN Reader r ON br.ReaderId = r.ReaderId
        WHERE 1=1";

            // 添加查询条件
            if (readerId.HasValue)
                sql += $" AND r.ReaderId = {readerId.Value}";

            if (!string.IsNullOrEmpty(readerName))
                sql += $" AND r.ReaderName LIKE '%{readerName}%'";

            if (!string.IsNullOrEmpty(bookName))
                sql += $" AND b.BookName LIKE '%{bookName}%'";

            if (!string.IsNullOrEmpty(isbn))
                sql += $" AND b.ISBN LIKE '%{isbn}%'";

            if (status.HasValue)
                sql += $" AND br.Status = {status.Value}";

            if (borrowFrom.HasValue)
                sql += $" AND br.BorrowDate >= '{borrowFrom.Value:yyyy-MM-dd}'";

            if (borrowTo.HasValue)
                sql += $" AND br.BorrowDate <= '{borrowTo.Value:yyyy-MM-dd}'";

            if (returnFrom.HasValue)
                sql += $" AND br.ReturnDate >= '{returnFrom.Value:yyyy-MM-dd}'";

            if (returnTo.HasValue)
                sql += $" AND br.ReturnDate <= '{returnTo.Value:yyyy-MM-dd}'";

            // 获取总记录数
            string countSql = "SELECT COUNT(*) FROM (" + sql + ") as t";
            object countResult = SQLHelper.ExecuteScalar(countSql);
            totalRecords = countResult != null ? Convert.ToInt32(countResult) : 0;

            // 添加排序和分页
            sql += " ORDER BY br.BorrowDate DESC";

            if (pageSize > 0)
            {
                int offset = (pageIndex - 1) * pageSize;
                sql += $" OFFSET {offset} ROWS FETCH NEXT {pageSize} ROWS ONLY";
            }

            DataSet ds = SQLHelper.ExecuteDataSet(sql);
            return ds?.Tables?[0];
        }

    }
}